App Permissions API Proposal

AydinHassan opened the pull request.

AydinHassan committed:
Add app permissions proposal

AydinHassan committed:
Update paths

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+            "get": {
+                "tags": ["App System"],
+                "summary": "Get requested privileges",
+                "description": "Returns the list of requested privileges for all apps.",
```
would be great to add the requirement as well
```suggestion
                "description": "Returns the list of requested privileges for all apps. Requires admin scope and `acl_role:read` permission to read.",
```

Ideally it would be best to add something referencing to `/token` endpoint and that `admin scope` is set there. Also where acl_role permission is set, so it's clear for the person invoking the endpoint what exactly needs to be done here

AydinHassan replied:
Do you have some example where we have done that previously ?

patzick replied:
I mean to add for example link to /token endpoint, where the available `scope`  list is visible, like https://shopware.stoplight.io/docs/admin-api/9384af2680768-fetch-an-access-token. As for the ACL - I don't know where is some docs about how to check and set it.

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+        "/app-system/privileges/requested": {
+            "get": {
+                "tags": ["App System"],
+                "summary": "Get requested privileges",
```
```suggestion
                "summary": "Get requested privileges from all apps.",
```

to be perfectly clear - is it like `pending` privileges or all ever requested? My guess is that it's a pending queue - would be nice to clarify it in naming

AydinHassan replied:
Yes it's just a pending queue 👍. I'll update.

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+                                },
+                                "example": {
+                                    "requestedPrivileges": {
+                                        "SwagAnalytics": {
+                                            "media": [
+                                                {"extensions": [], "entity": "media", "operation": "read"}
+                                            ],
+                                            "settings": [
+                                                {"extensions": [], "entity": "state_machine", "operation": "read"},
+                                                {"extensions": [], "entity": "state_machine_state", "operation": "read"}
+                                            ]
+                                        }
```
the object is very generic and it's not sure what types of fields should we expect in it, would be nice to dig deeper into what are the possibilities for these fields and available options

Additionally - how would `extensions` work in this case?

AydinHassan replied:
Do you mean just the extensions bit or? How would you type it? The extensions bit i'll dig into, it's from another API with similar payload that is untyped.

patzick replied:
Yeah, I mean what would be the purpose of that? It's not clear for now :)

So What is also missing are "possible values", so:
- App name is a string - that we know and that's fine, but do we ensure there are no duplicates in app names?
- What are possible "group names"
- what are possible "entities"
- what are possible "operations"

Schema should define that all, so it's perfectly clear what can we expect here. I'd also consider also `description` fields or something - how user can cauciasly grant access to "state_machine:read"? What impact does it have? What app will be actually allowed to do with that? Displaying description would help makind proper decisions

Gaitholabi replied:
I will look this up, not too sure about those payloads either

Gaitholabi replied:
The 'extensions' aren't very relevant, so I've removed them now. The APIs are simplified now.

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+                            }
+                        }
+                    },
+                    "500": {
+                        "description": "Incorrect source. "
+                    }
```
I believe we should not return 500 "by design". Also - what would "incorrect source" mean?

My propositions:
- 401 - when not authenticated
- 403 - no access to this endpoint, explaining in the description that it means that scope or ACL is not correct

AydinHassan replied:
Sure - I wanted feedback here, I did not dig too much into our errors in other schemas. I'll update based on your suggestions :)

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+                }
+            }
+        },
+        "/app-system/privileges/accepted" : {
```
Isn't that more an "audit" endpoint where you could review all applications granted permissions? So practically the same as `/app-system/privileges/requested`

That's why I would combine them as one `GET /app-system/privileges` for clarity and then `GET /app-system/{appName}/privileges` for the App to get its context

The body could share the same component and dictionaries as other endpoints

MartinKrzykawski replied:
I like the idea! Merging the `requested` and `accepted` endpoints into a single, streamlined listing endpoint would create a more intuitive structure, especially with the use of pluralized resource names. Additionally, we could introduce filtering options to allow fetching privileges based on specific criteria, such as status.

AydinHassan replied:
No - it's so the app backend can fetch its currently approved set of permissions, so it can store them/act on them. They will also receive webhooks when they are changed so the end point should not be used often, just rather as a backup. 

Right now i would really prefer not to combine them because we already have end points with similar payloads and this will make the client code much more complicated, but I also see we want to design proper API's so I will have a think about it. 

We don't really need the app to know about requested permissions, and in that case I would prefer to expose less data. Only the admin needs to know about requested permissions, and that's for all apps, not the just the current app like here.

patzick replied:
I left one of my biggest comment in the "review" not here as comment to line of codes, so might have been missed, related: https://github.com/shopware/api-working-group/pull/1#pullrequestreview-2718236840

I proposed 3 endpoints there in total, so for an App to get it's granted privileges I'd say that should be in the app namespace:
`GET /api/app-system/{appName}/privileges` - this way it's clear that this returns privileges only for specific app and it should be used by this app to fetch it. Even if hat's only a "backup" endpoint for an app

Gaitholabi replied:
I think this conversation is also extended here: https://github.com/shopware/api-working-group/pull/1#discussion_r2027343844

I'll resolve in favor of continuing there.

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+                    },
+                    "500": {
+                        "description": "Incorrect source. "
+                    }
```
here the same with proposed responses
- 401
- 403
- 404 (optionally?) - when app is not installed? Or that would fail sooner with 403?

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+                                    "type": "string"
+                                }
+                            },
+                            "example": ["customer:read", "order:read"]
```
how would it be from the example from above?
```
"customer": [
      {"extensions": [], "entity": "customer", "operation": "read"},
      {"extensions": [], "entity": "customer_group", "operation": "read"}
  ],
```
then we should send `customer:customer_group:read` ? How's the grouping work, as in most examples the group name is repeated as entity name

AydinHassan replied:
It would be `["customer_group:read", "customer:read"]` - there is no grouping, they are separate entities in the system.

patzick reviewed (commented):
Thank you for first API proposal 🎉 🙌 


What I see here is 2 endpoints for the administration of the Apps, and 1 endpoint for the App to get its permissions. What I miss is:
- ability to reject privilege
- ability to easily "sort" the data by permission not app, to autit which apps have access to specific resource
- ability to revoke privilege

As for the more payload point of view:
- not super clear payload structure (more in comments), extracting common schemas as components
- adding dictionaries, for that I might not have enough context so happy to discuss further

## Proposed changes
So taking a grasp of what do we have two perspective

### Managing privileges
- `GET /api/app-system/privileges` - where we could have permissions with the statuses `requested | accepted | rejected` - so in administration view it could be used for purposes of auditing states and getting the requested ones. I added `rejected` state so it would not come up after every update and user can just decline it
- `POST /api/app-system/privileges` - sending a list of updates containing: appName, privilege status and specifics about the entity (again need more context with possible dictionaries)

### Getting Granted Privileges
As an app (so in the App namespace) we just need to get the info:
- `GET /api/app-system/{appName}/privileges` - this could return also privileges with status so app even know if specific privilege was rejected or hasn't been considered yet

MartinKrzykawski commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,207 @@
+
+### API Endpoints
+
+*/api/app-system/privileges/requested*
```
I believe using nouns to represent resources instead of verbs aligns better with REST API naming conventions and best practices: https://medium.com/@nadinCodeHat/rest-api-naming-conventions-and-best-practices-1c4e781eb6a5

Here’s my suggestion:

For privileges within a single app:
* `GET /app-system/{appName}/privileges` to retrieve privileges in the app context.
* `PUT /app-system/{appName}/privileges` to update privileges or change their status within the app context. This would make the accept endpoint obsolete and provide a unified approach.

For all privileges:
* `GET /app-system/privileges` to get a list of all privileges.
* `GET /app-system/privileges?status=requested&limit=10` - using filters like "status" or "limit" would offer more flexibility and control over the data merchants can retrieve.
* `PUT /app-system/privileges` to update privileges and change their status in a single request. For example:
```
{
   "apps": [
      {
         "appName": "app1",
         "privileges": [
            { "id": "123", "status": "accepted" },
            { "id": "456", "status": "rejected" }
         ]
      },
      {
         "appName": "app2",
         "privileges": [
            { "id": "789", "status": "pending" },
            { "id": "101", "status": "accepted" }
         ]
      }
   ]
}
```

We could also use PATCH instead of PUT for partial updates, if only certain attributes (like status) need to be modified.

AydinHassan replied:
We had a discussion about using the nouns elsewhere and rejected it. I'll find it and cross link. Regarding the statuses/filtering etc I would rather not blow this tiny feature out of the water (these payloads will be rather small and not oft used) but I'll have a think about it based on all the feedback here :)

AydinHassan replied:
Hmm - I think I was mistaken - it was something else. Regardless, I don't reallly think it fits here. We don't internally have or want the concept of permission statuses. It's either pending or accepted. There is no rejected status. That would in itself be a huge topic with many considerations and I don't believe we should pave the way for the unknown.

patzick replied:
Also relevant for this thread: https://github.com/shopware/api-working-group/pull/1#pullrequestreview-2718236840

Even if the status is not something we consider at this point, it would be still good to design process to revoke permission. What should I do if i granted permission by mistake? Do I need to uninstall the app?

So if we already have two groups:
- requested
- accepted

Then `rejected` status doesn't need to exist, revoked permission can go back to `requested` and that's it.

But from the managing perspective you still need to invoke two endpoints to get all the data you need to see, that's why `status` make sense for the presentation layer, even if that's not a thing in the system

then endpoint
`GET /api/app-system/privileges` would return everything that app permission manager needs to know - both accepted and requested permissions to switch this "abstract status" between them

keulinho replied:
I would keep the two endpoints separate and not add the status to them, because:
* to get already accepted permission of apps there are already endpoints and we already use them, no need to reinvent the wheel and overcomplicate the system, that way we can add the requested state without having to rework everything that touches app permissions
* the clients for the requested permissions and accepted privileges are totally different, therefore IMHO it does not make sense to combine them to use one endpoint, that will only make the clients more complicated

The requested permissions are used from the administration when we ask for giving those permissions, the already given permissions are not relevant at that time (and if there are already enpoints to get those that are already used in other places) 

The accepted permissions are fetched from the app backend to get the current state, the app is not interested at all in the requested permissions, as the app knows which permissions it requested, it needs to know which permissions it was actually granted

therefore I like the proposal and the split of the API endpoints for both use cases, for me it's a prime example of thinking API endpoints from the client first and not overgeneralize some API endpoint that could be used for everything and then is complicated to use for any real usecase.

One note though: I also think naming of both endpoints is a bit confusing and probably let to a lot of those discussions as the `/api/app-system/privileges/requested` and `/api/app-system/privileges/accepted` sound like they are basically the same, or work in the same way because the naming is similar

could we solve this by renaming the `accepted` route to `/api/app-system/my-privileges` or something similar that makes clear that the intended use case is from the app and that the response is scoped to the app that requests it? That is way easier to use for apps then to add the app name to the URL, as that implies that you could also fetch the permissions of other apps, but that is not the case

patzick replied:
> to get already accepted permission of apps there are already endpoints 

Could we put these endpoints here for the context? That would also help with the consistency of the responses

> the clients for the requested permissions and accepted privileges are totally different

That's interesting perspective. I'd say the client is the same (app manager who wants to manage priviledges), but the view is different, and the views changes often. From my perspective it's just splitting filtering in two endpoints, so for example the view of "App permissions" would need to fetch the data from two sources.

That said it's fine to have it like this (2 endpoints) if the payloads contains the same structure, it's described well, predictable and understandable.

So in summary:
1. @keulinho - You mentioned there are already some endpoints to get accepted permissions - can you link them here?
2. Would be nice to have summary of what we already have vs what we need, as I understood we needed:
  a. [Manager/Admin] review requests for all apps 
  b. [Manager/Admin] accept requests
  c. [App] getting its accepted privileges (so this one already exist?)
4. I'm still wandering what about revoking permissions, which is completly valid case

As for the last proposal, I do not see `(GET/POST?) /api/app-system/my-privileges` being more clear than `GET /api/app-system/{appName}/privileges` to get accepted privileges for specific application. The AppName namespace is suggesting very clearly that this resource is for the specific app

Gaitholabi replied:
hey, I'm reviving this topic in order to move the API further. I partially agree with everyone, and suggest the following:
1. Rename: `/api/app-system/privileges/accepted` -> `/api/app-system/{appName}/privileges/accepted`

The rename here signifies that this is always called with the context of the integration, as I initially thought that it was also global for all apps.

2. Rename and change method: POST `/app-system/{appName}/privileges/accept` -> PATCH `/app-system/{appName}/privileges/`

This makes it more flexible, accommodating for accepting and revocation since we need the two functionalities. The body would look something like:
```json
{
    "accept": [....],
    "revoke": [.....]
}
```

@keulinho I quickly skimmed the API docs, but cannot find APIs returning information about app permissions 🤔

Gaitholabi replied:
@patzick do you think that what I described also fits with your suggestions?

patzick replied:
hey @Gaitholabi thanks for taking care of the topic further 🙌 

as for your input I'm good with this outcome. We have here multiple threads so would be super nice to gather it all, update the RFC proposal and do the second round of the reviews.
For example you haven't mentioned `GET /app-system/privileges/requested`, so not sure if that's already okayed or just missed in the summary (I assume first option) :)

With all updated data and clearly defined schemas so then we can accept it and move to the implementation :)

Gaitholabi replied:
For me, I agreed in full about `GET /app-system/privileges/requested`, so I there was no change here. 

I'll update the working examples 👍

Gaitholabi replied:
hey @keulinho @patzick,
I've reworked the suggestion a bit, if you could take another look please, and here's a summary of the changes:

note: some of the below are mentioned in the previous comment.

- `GET /app-system/privileges/requested`
the response has been simplified, and repurposed. Reason: remove entities and the exposure of Shopware internals (entity, categories..etc) details from response, those are irrelevant for the API purpose. 

Rename: `GET /api/app-system/privileges/accepted` -> `GET /api/app-system/{appName}/privileges/accepted`
The request now makes the app context clearer. Response is now more tailored towards clients ease of parsing, and processing. in the context of this response, clients would most likely want to see if some permission is there before it executes 

- Rename and change method: `POST /app-system/{appName}/privileges/accept` -> `PATCH /app-system/{appName}/privileges/`

This makes makes the API more flexible, and still keeps the functionality scoped to the same concern.

MartinKrzykawski replied:
Thanks for taking care! :) I am very happy with the result. I found the different perspectives throughout the discussion really insightful, and I think the final version strikes a great balance and the changes bring more clarity to the API while also making it more flexible and maintainable. Great work! 🚀

Gaitholabi committed:
rework granting endpoint and make it PATCH

add app name parameter to permission accepted context

remove 500 statuses

convert the schema to YAML

Gaitholabi committed:
convert back to json

Gaitholabi committed:
simplify body and schema

MartinKrzykawski commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,136 @@
+
+Fetch the set of accepted permissions for the current integration. Requires admin scope with an integration.
+
+### Schema
```
The schema just needs to be updated to reflect the latest outcome — other than that, everything looks great imo! 👍

Gaitholabi replied:
I already updated it, can you please point me where do you see the inconsistency? 🤔

Gaitholabi replied:
I see, my AI forgot the patch endpoint all together 💀

Gaitholabi committed:
docs: complete missing patch endpoint

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
```diff
@@ -0,0 +1,241 @@
+
+Fetch the set of requested permissions. It retrieves all requested permissions for all apps. Requires admin scope and `acl_role:read` permission to read. 
+
+*/api/app-system/{appName}/privileges/accept* 
```
this one is incorrect I think - should be `PATCH /app-system/{appName}/privileges`

patzick commented on design-decisions/rest-admin-api/2025-03-18-appPermissions.md:
````diff
@@ -0,0 +1,241 @@
+
+```json
+{
+  "openapi": "3.0.0",
````
```suggestion
  "openapi": "3.1.0",
```
we use 3.1.0

patzick reviewed (dismissed):
It looks alright, I just found out one very small inconsistency but we can merge it after so I'm leaving the accept already. Very nice collaboration here 🙌

Gaitholabi committed:
fix review comments and inconsistency
